MapboxCoreNavigation/RouteController.swift

```diff
@@ -342,36 +342,53 @@ open class RouteController: NSObject {
             let finalHeading = upcomingStep.finalHeading,
             let coordinates = routeProgress.currentLegProgress.currentStep.coordinates {
             
-            // Calculate if angle is sharp
-            let inAngle: CLLocationDegrees = initialHeading.toRadians()
-            let outAngle: CLLocationDegrees = finalHeading.toRadians()
-            
-            let inX = sin(inAngle)
-            let inY = cos(inAngle)
-            let outX = sin(outAngle)
-            let outY = cos(outAngle)
-            
-            let turnAngle = acos((inX * outX + inY * outY) / 1.0) * (180 / .pi)
-            
             // The max here is 180. The closer it is to 180, the sharper the turn.
-            if turnAngle > 180 - RouteControllerMaxManipulatedCourseAngle {
+            if turnAngleStraightness(angleIn: initialHeading, angleOut: finalHeading) > 180 - RouteControllerMaxManipulatedCourseAngle {
                 nearByCoordinates = coordinates
             }
         }
         
-        let nearByPolyline = Polyline(nearByCoordinates)
-
         guard let closest = Polyline(nearByCoordinates).closestCoordinate(to: location.coordinate) else { return nil }
-
-        let slicedLineBehind = Polyline(nearByCoordinates.reversed()).sliced(from: closest.coordinate, to: nearByCoordinates.reversed().last)
-        let slicedLineInFront = Polyline(nearByCoordinates).sliced(from: closest.coordinate, to: nearByCoordinates.last)
+        guard let calculatedCourseForLocationOnStep = interpolatedCourse(from: location, along: nearByCoordinates) else { return nil }
+        
+        var userCourse = calculatedCourseForLocationOnStep
+        var userCoordinate = closest.coordinate
+        if shouldAllowUnsapped(location: location, given: calculatedCourseForLocationOnStep) {
+            userCourse = location.course
+            userCoordinate = location.coordinate
+        }
+        
+        return CLLocation(coordinate: userCoordinate, altitude: location.altitude, horizontalAccuracy: location.horizontalAccuracy, verticalAccuracy: location.verticalAccuracy, course: userCourse, speed: location.speed, timestamp: location.timestamp)
+    }
+    
+    func turnAngleStraightness(angleIn: CLLocationDirection, angleOut: CLLocationDirection) -> CLLocationDirection {
```

Comment:
This would be a good candidate for upstreaming to Turf. As I understand it, the only difference between this method and `CLLocationDirection.difference(from:)` is that the latter takes the short way around, always returning a positive value.

Regardless, let’s extract this method from RouteController, because it doesn’t have any dependencies on `self`. Put it in it an extension of `CLLocationDirection`, as an instance method named `clockwiseDifference(from:)` – assuming it actually does measure clockwise, as implied by `CLLocationDirection`.